### ISSUE TITLE
Shorten long URLs

### ISSUE THREAD

**Issue by rqu1** [tuskyapp/Tusky#655]
The web client visually truncates URLs to 30 characters, which makes scrolling past [very long links](https://mastodon.social/@scanlime/100053418346841470) not take forever. Proposed new behavior would be to copy the web clients behavior and truncate link text after a few characters instead of displaying the full URL.

* * * *
- Tusky Version: 1.8.0
- Android Version: 8.0.0
- Android Device: Samsung Galaxy s9

- [X] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by charlag** [comment]
Hi! Thanks for reporting!

**Comment by connyduck** [comment]
I actually think Mastodon shortens urls too soon. You cant even see if a github link is an issue or a pull request. On the desktop its possible to preview the whole link by hovering, we don't have that on mobile,
But yeah we should do something about really long urls, I think truncating them at 50-60 would be ok?

**Comment by pdjstone** [comment]
I had a look at the code in `LinkHelper.kt`, and I think this should be fairly easy to implement. This is my first time looking at the code, but I think you would just need another function, called from `setClickableText` that iterated over the URLSpans, and replaces the link text for long URLs with a shortened version.

I can have a go at implementing this, but I have a few questions:

- Do we also want to hide the protocol (http/https) like Mastodon web does?
- Do we want to include a way to preview the full link (e.g. by long pressing the link), similar to how Chrome or Firefox does?

**Comment by connyduck** [comment]
Actually the HTML Mastodon sends us already has the urls shortened, but the HTML parser we use does not handle it correctly.

`<p>Volkswagen: all brands have halted paid activities on Twitter <a href=\"http://reut.rs/3EY8h2n\" rel=\"nofollow noopener noreferrer\" target=\"_blank\"><span class=\"invisible\">http://</span><span class=\"\">reut.rs/3EY8h2n</span><span class=\"invisible\"></span></a></p>`

Maybe hooking into the HTML (with a custom [`TagHandler`](https://developer.android.com/reference/android/text/Html.TagHandler)) is a better idea than implementing our own url shortening?

To answer the question: I like how Mastodon web shortens urls, but it might not be ideal for touchscreens (can't hover to preview), so yes long press to preview sounds like a good idea.

**Comment by pdjstone** [comment]
I did experiment with a TagHandler in `parseAsMastodonHtml` however because `HtmlCompat.fromHTML` treats a span as a supported tag, we can't modify it that way. 

You mentioned above that Mastodon perhaps shortens paths too much - if the Tusky code does shortening itself, we have more control over that. I have some code and tests that are working for the shortening. When I get some time I'll have a go at the dialog.

**Comment by pdjstone** [comment]
I realised that by shortening links inside of `setClickableText`, the overall status length changes compared to `StatusViewData.content`. This causes the calculation for `StatusViewData.isCollapsible` to be incorrect. As a result, the collapse button is sometimes shown when it shouldn't be, and in the wrong place.

My WIP code is in commit ac484f03 but I need to have a think about how to refactor things to handle this correctly.

**Comment by nikclayton** [comment]
https://mstdn.ca/@fishgolfski/109518030131969248 is a user request for this feature.

**Comment by TodePond** [comment]
Hi just to come and chime in here, this seems pretty important for people sharing URLs with long queries, which is very common in the live coding / demoscene. eg: https://mas.to/@TodePond/114160164139319708

Nearly all other clients handle them well.

eg: Elk

![Image](https://github.com/user-attachments/assets/3802e8a2-0cb4-43cf-84f2-bd16c1207ce2)

eg: Default mastodon web client

![Image](https://github.com/user-attachments/assets/983bb092-c0d1-44d1-bab3-2873130d2b8e)

But unfortunately, Tusky gives people a bad time.

https://github.com/user-attachments/assets/bacfb162-8d2f-4304-82c9-5474a5bb321c

Just raising this in case anyone is looking for some low-hanging-fruit that would be high impact :)

**Comment by connyduck** [comment]
If we shorten urls we definitely need a way to preview them before clicking. Long press probably.

**Comment by TodePond** [comment]
> If we shorten urls we definitely need a way to preview them before clicking. Long press probably.

I don't think you would? Most other clients dont seem to do this. Long press sometimes copies the URL i think

**Comment by TodePond** [comment]
> It is important for security purposes though. Wouldn't want to click on a link I don't know if I can't see it fully. On desktop that is easier as one can hover links.

I really think this is a nice-to-have, not a necessary thing. That's really not the expectation with links on the fediverse. Yet again, people have had this issue.

https://elk.zone/mas.to/@jackie@deadinsi.de/114199805576328910
https://elk.zone/xoxo.zone/@annika/114160264816372883

**Comment by Tak** [comment]
imo it's *also* not the expectation with links on the fediverse that they're thousands of characters long
while it will be nice to support link shortening, it is *absolutely* not worth dropping a security feature to do so

**Comment by TodePond** [comment]
That's a shame, its quite common with sharing live coding + demo scene stuff. I'll keep sending people here

**Comment by Tak** [comment]
It's fine, we'll do it when someone has time and inclination, we're just not going to take a quick shortcut 🤷
